### PR TITLE
Framework to aid in constructing JDBC datasources

### PR DESCRIPTION
**Pull request by wemrysi**
Beginnings of a framework for building JDBC quasar plugins, initially focused on datasources.

[ch11225]

### CODE REVIEW

File: core/src/main/scala/quasar/plugin/jdbc/JdbcDiscovery.scala [hunk 1]
```diff
@@ -0,0 +1,234 @@
...
+import cats.data.{Ior, NonEmptySet}
+import cats.implicits._
+
+import doobie._
+import doobie.enum.{ColumnNullable, JdbcType}
+import doobie.implicits._
+
+import fs2.Stream
+
+/** Provides a means of discovering schemas and tables in a database. */
```
Comment (alissapajer) [inline core/src/main/scala/quasar/plugin/jdbc/JdbcDiscovery.scala:35]: Regarding all the schema-related functionality introduced here: do/will we expose access to tables via database schema in Precog? Or is the schema functionality not going to be used in production yet?

File: core/src/main/scala/quasar/plugin/jdbc/package.scala [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+
+  val Redacted: String = "<REDACTED>"
+
+  def resourcePathRef(p: ResourcePath): Option[DboRef] =
+    Some(p) collect {
+      case fst /: ResourcePath.Root =>
+        Left(Ident(fst))
+
+      case fst /: snd /: ResourcePath.Root =>
+        Right((Ident(fst), Ident(snd)))
```
Comment (alissapajer) [inline core/src/main/scala/quasar/plugin/jdbc/package.scala:51]: This answers my question about how schema are used, at least partly.

File: core/src/main/scala/quasar/plugin/jdbc/datasource/JdbcDatasource.scala [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+
+final class JdbcDatasource[F[_]: Bracket[?[_], Throwable]: Defer] private (
+    xa: Transactor[F],
+    discovery: JdbcDiscovery,
+    val kind: DatasourceType,
+    val loaders: NonEmptyList[Loader[Resource[F, ?], InterpretedRead[ResourcePath], QueryResult[F]]])
+    extends LightweightDatasourceModule.DS[F] {
+
+  def pathIsResource(path: ResourcePath): Resource[F, Boolean] =
+    resourcePathRef(path).fold(false.pure[Resource[F, ?]]) {
```
Comment (alissapajer) [inline core/src/main/scala/quasar/plugin/jdbc/datasource/JdbcDatasource.scala:53]: More specifically, here, if the `path` contains a schema then we'll discover the table with the given name in that schema.

File: core/src/main/scala/quasar/plugin/jdbc/datasource/JdbcDatasource.scala [hunk 4]
```diff
@@ -0,0 +1,111 @@
...
+        case Right((schema, table)) =>
+          Resource liftF {
+            discovery.tableExists(table, Some(schema))
+              .map(p => if (p) Some(Stream.empty: Out[F]) else None)
+              .transact(xa)
+          }
+
+        case Left(ident) =>
+          def paths =
+            discovery.tables(Ior.left(ident))
```
Comment (alissapajer) [inline core/src/main/scala/quasar/plugin/jdbc/datasource/JdbcDatasource.scala:83]: Here, the `Left(ident)` uses `ident` as a schema identifier. But up on line 54 in this file, `Left(table)` uses `table` as a table identifier. Shouldn't `resourcePathRef` behave consistently?

Comment (alissapajer) [reply core/src/main/scala/quasar/plugin/jdbc/datasource/JdbcDatasource.scala:83]: Do we want `Ior.right(ident)` here instead?

File: core/src/main/scala/quasar/plugin/jdbc/JdbcDiscovery.scala (hunk 1 above)
Comment (wemrysi) [reply core/src/main/scala/quasar/plugin/jdbc/JdbcDiscovery.scala:35]: @alissapajer It is in the path, we expose the database objects as `/<schema>/<table>`

File: core/src/main/scala/quasar/plugin/jdbc/datasource/JdbcDatasource.scala (hunk 4 above)
Comment (alissapajer) [reply core/src/main/scala/quasar/plugin/jdbc/datasource/JdbcDatasource.scala:83]: The larger question really is: when `resourcePathRef` returns a single value, (not a tuple), how do we know if that value is the schema name or the table name?

File: core/src/main/scala/quasar/plugin/jdbc/JdbcDiscovery.scala (hunk 1 above)
Comment (wemrysi) [reply core/src/main/scala/quasar/plugin/jdbc/JdbcDiscovery.scala:35]: Most of this was cribbed from the Postgres datasource and generalized

File: core/src/main/scala/quasar/plugin/jdbc/datasource/JdbcDatasource.scala [hunk 5]
```diff
@@ -0,0 +1,111 @@
...
+              .map(m => (ResourceName(m.table.asString), RPT.leafResource))
+              .pull.peek1
+              .flatMap(t => Pull.output1(t.map(_._2)))
+              .stream
+
+          for {
+            c <- xa.strategicConnection
+
+            isTable <- Resource.liftF(xa.runWith(c).apply(discovery.tableExists(ident, None)))
```
Comment (alissapajer) [inline core/src/main/scala/quasar/plugin/jdbc/datasource/JdbcDatasource.scala:92]: Here `ident` is used as the table name again.

File: core/src/main/scala/quasar/plugin/jdbc/datasource/JdbcDatasourceModule.scala [hunk 6]
```diff
@@ -0,0 +1,120 @@
...
+
+import org.slf4s.{Logger, LoggerFactory}
+
+/** A Quasar LightweightDatsourceModule for JDBC sources.
+  *
+  * Handles boilerplate common to all JDBC datasources, such as
+  *   - parsing JSON into a vendor-specific config
+  *   - constructing a pooled `Transactor` along with the necessary threadpools
+  *   - validating a connection to the database can be established
+  *   - logging
```
Comment (alissapajer) [inline core/src/main/scala/quasar/plugin/jdbc/datasource/JdbcDatasourceModule.scala:49]: 🎉

File: core/src/main/scala/quasar/plugin/jdbc/datasource/JdbcDatasourceModule.scala [hunk 7]
```diff
@@ -0,0 +1,120 @@
...
+  */
+abstract class JdbcDatasourceModule[C: DecodeJson] extends LightweightDatasourceModule {
```
Comment (alissapajer) [inline core/src/main/scala/quasar/plugin/jdbc/datasource/JdbcDatasourceModule.scala:51]: Will we be able to (or want to) port any of our existing datasources over to use this?

File: core/src/main/scala/quasar/plugin/jdbc/datasource/JdbcDatasource.scala (hunk 4 above)
Comment (wemrysi) [reply core/src/main/scala/quasar/plugin/jdbc/datasource/JdbcDatasource.scala:83]: I named it `ident` to convey the ambiguity, as you point out, we don't yet know if it refers to a schema or a table, it is just an identifier.

> The larger question really is: when resourcePathRef returns a single value, (not a tuple), how do we know if that value is the schema name or the table name?

Indeed, this is handled below, we first check if `ident` is a table without a schema, if so we're done, else we treat it as a schema and return `paths`.

File: core/src/main/scala/quasar/plugin/jdbc/datasource/JdbcDatasourceModule.scala (hunk 7 above)
Comment (wemrysi) [reply core/src/main/scala/quasar/plugin/jdbc/datasource/JdbcDatasourceModule.scala:51]: Yes, ideally all of them that use JDBC. We'll also want to add support for destinations and convert those as well.

File: core/src/main/scala/quasar/plugin/jdbc/datasource/MaskedScalarColumns.scala [hunk 8]
```diff
@@ -0,0 +1,74 @@
...
+      }
+
+      scalars match {
+        case h :: t => isDefined(c, NonEmptySet.of(h, t: _*))
+        case Nil => false
+      }
+    }
+
+    scalarStages match {
+      case ScalarStages(excludeId @ IdStatus.ExcludeId, ScalarStage.Mask(m) :: rest) =>
```
Comment (alissapajer) [inline core/src/main/scala/quasar/plugin/jdbc/datasource/MaskedScalarColumns.scala:54]: Ok, it's _really_ cool that we can so easily push down a `ScalarStage` evaluation to a datasource like this.

File: core/src/main/scala/quasar/plugin/jdbc/datasource/JdbcDatasource.scala (hunk 4 above)
Comment (alissapajer) [reply core/src/main/scala/quasar/plugin/jdbc/datasource/JdbcDatasource.scala:83]: Ohh I see how this is working now! `paths` above is only actually used when we land in the `!isTable` case below. All good! :)

**Comment by alissapajer** [comment]
I'm super excited to have this plugin generalized in this way!

Would we be able to test any more of this similar to how `doobie` has tested itself internally, [acquiring a transactor from h2](https://github.com/tpolecat/doobie/blob/master/modules/core/src/test/scala/doobie/util/GetSpec.scala#L65-L69)?

**Comment by alissapajer** [comment]
@wemrysi could you add some info to the README, a broad strokes "how to use this to implement an actual plugin" sort of thing. Or whatever would be useful to someone who wants to use this.

**Comment by wemrysi** [comment]
@alissapajer 

> Would we be able to test any more of this similar to how doobie has tested itself internally, acquiring a transactor from h2?

Yes, I plan to at least have some general tests using `h2`.

For timing reasons, I'd prefer to backfill those once the Avalanche source is done.

> Could you add some info to the README, a broad strokes "how to use this to implement an actual plugin" sort of thing.

Yes, though I expect things to evolve a bit once we add destination support (I'm not sure this is _quite_ ready for anyone to pick up and use). So I'll keep things fairly broad and write up more detailed docs once things look stable.

**Review by alissapajer — APPROVED** [review]

**Comment by alissapajer** [comment]
🎉

**Pull request merged**